Fusion layer instantiates parts lazily

tinwelint opened the pull request:
Instead of always instantiating eagerly. E.g. an IndexReader would previously
instantiate sub-readers for lucene, number, string, spatial and temporal
even if it would only need one of those. Likewise they would all be closed
afterwards. This proved to be an unnecessary overhead.

tinwelint committed:
Fusion layer instantiates parts lazily

Instead of always instantiating eagerly. E.g. an IndexReader would previously
instantiate sub-readers for lucene, number, string, spatial and temporal
even if it would only need one of those. Likewise they would all be closed
afterwards. This proved to be an unnecessary overhead.

tinwelint commented:
Marking as not ready since may skip GA and instead go into a patch release.

tinwelint committed:
Testing for the lazy instantiation in fusion layer

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/Selector.java:
```diff
@@ -0,0 +1,73 @@
+ * Selects an instance given a certain slot.
+ * @param <T> type of instance
+ */
+class Selector<T>
```
As we also have the `SlotSelector`, I think it will be more clear if this was called `InstanceSelector`.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexBase.java:
```diff
@@ -133,24 +162,22 @@
      * @throws E if consumption fails with this exception
      */
-    public static <T, E extends Exception> void forAll( ThrowingConsumer<T,E> consumer, Iterable<T> subjects ) throws E
+    public static <T, E extends Exception> void forAll( ThrowingConsumer<T,E> consumer, T[] subjects ) throws E
```
Update javadoc with reference to `forAll( ThrowingConsumer<T,E> consumer, Iterable<T> subjects )` and update javadoc for `@param subjects`.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexProvider.java:
```diff
@@ -40,47 +39,30 @@
-    }
 
     private final boolean archiveFailedIndex;
     private final IndexProvider[] providers = new IndexProvider[INSTANCE_COUNT];
```
We can get rid of this field now

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexBase.java:
```diff
@@ -133,24 +162,22 @@
+        return exception;
     }
 
     static void validateSelectorInstances( Object[] instances, int... aliveIndex )
```
This method should probably live with `SlotSelector`

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexProvider.java:
```diff
@@ -40,47 +39,30 @@
     private final boolean archiveFailedIndex;
     private final IndexProvider[] providers = new IndexProvider[INSTANCE_COUNT];
-    private final Selector selector;
+    private final Selector<IndexProvider> selector;
```
When we get rid of field `providers` we can use that name here instead, more clear.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexProvider.java:
```diff
@@ -180,11 +164,8 @@ public InternalIndexState getInitialState( long indexId, SchemaIndexDescriptor d
-            capabilities[i] = providers[i].getCapability( schemaIndexDescriptor );
-        }
+        IndexCapability[] capabilities =
+                instancesAs( selector, new IndexCapability[providers.length], provider -> provider.getCapability( schemaIndexDescriptor ) );
```
In other places we use `INSTANCE_COUNT` for array length. Could we do that here to?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/Selector.java:
```diff
@@ -0,0 +1,73 @@
+            catch ( Exception e )
+            {
+                // TODO figure out how to handle this
+                throw new RuntimeException( e );
```
Yes, indeed

tinwelint replied:
Solved by avoiding `ThrowingIntFunction` altogether. It was there only because of the `IndexPopulator#newPopulatingUpdater()` throwing `IOException`, but apparently no implementation did so, so removed it from the signature.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:
```diff
@@ -122,6 +117,5 @@ public boolean hasFullValuePrecision( IndexQuery... predicates )
             // full value precision for that, therefor true.
             return true;
```
Remove or update this comment

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessorTest.java:
```diff
@@ -131,7 +131,9 @@ private void initiateMocks()
-        fusionIndexAccessor = new FusionIndexAccessor( accessors, fusionVersion.selector(), indexId, mock( SchemaIndexDescriptor.class ), dropAction );
+        fusionIndexAccessor =
+                new FusionIndexAccessor( fusionVersion.slotSelector(), new Selector<>( accessors ), indexId, mock( SchemaIndexDescriptor.class ),
+                        dropAction );
```
Can live on fewer lines

tinwelint replied:
IntelliJ multi-line formatting is very rarely optimal

burqen replied:
Problem is that it does not remove new lines, only add.

burqen reviewed (changes requested):
Got some comments and one point where readability could be improved:
We now use `Selector` everywhere where we used to use `T[]`. This means that all places where we delegate downwards from Fusion are potentially lazy. It would be nice if we only used Selector where we actually need instantiation to be lazy and something more explicit where we don't.
Do you think this is feasible ?

tinwelint committed:
Better readability and naming of fusion selector interfaces

tinwelint commented:
About going for T[] sometimes... the way I look at is that it would be for performance, otherwise using the InstanceSelector in all places makes it easier to understand the code IMHO. The places where the instances are instantiated eagerly are places that don't have performance implications, so I'd rather leave it like this w/ selector everywhere.

tinwelint commented:
@burqen updated

burqen commented:
@tinwelint Nice cleanup! As per offline conversation: Add `close`, `forAll`, `instancesAs` and `forInstantiated` to `InstanceSelector`. This will leave the static methods in `FusionIndexBase` only used by Spatial and Temporal and could then live closer to them.

tinwelint committed:
Moved methods for sub-instances to InstanceSelector

Instead of them being static in FusionIndexBase.
InstanceSelector also got a close method which will prevent
lazy instantiation after being closed.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/InstanceSelector.java:
```diff
@@ -51,17 +58,137 @@
         return instances[slot];
     }
 
+    private void assertOpen()
```
This guy can live at the bottom with other private methods.

tinwelint replied:
You're bike shedding slightly

burqen replied:
True!

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/InstanceSelector.java:
```diff
@@ -51,17 +58,137 @@
         if ( instances[slot] == null )
         {
+            assertOpen();
             instances[slot] = factory.apply( slot );
```
I think we have a race here with close.
```
T1 check assertOpen
T2 close()
T1 factory.apply( slot )
```

tinwelint replied:
Selectors, IndexUpdaters and IndexReaders aren't meant to be thread safe.

tinwelint commented:
@burqen updated

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/IndexPartsCache.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+abstract class IndexPartsCache<KEY,T> implements Iterable<T>
```
I'm not sure I like this breakout. For two reasons:
1. I think it's outside the scope of this PR.
2. It feels like there are much more that should be done to clean this part of the code up and to me this "half way" change makes it even harder to understand because of the direct access of lock and cache in subclasses.|

My suggestion is that we do a proper refactoring after GA.

tinwelint replied:
Did it because there was one name change related to a previous commit in this PR that was previously only done in one of the two duplicated classes, but forgotten in the other. The duplication is silly and I don't think we're off worse with this change. If you really think we're in a worse position now then I can revert it.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/InstanceSelector.java:
```diff
@@ -51,17 +58,137 @@
+     * @return the target array which was passed in, now populated.
+     * @throws E exception from converter.
+     */
+    <R,E extends Exception> R[] instancesAs( R[] target, ThrowingFunction<T,R,E> converter ) throws E
```
Moving all of those methods here from `FusionIndexBase` made things so much clearer! 🎉

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessorTest.java:
```diff
@@ -251,20 +251,22 @@ public void closeMustCloseAll() throws Exception
     public void closeMustThrowIfOneThrow() throws Exception
     {
-        for ( IndexAccessor accessor : aliveAccessors )
+        for ( int i = 0; i < aliveAccessors.length; i++ )
```
Why change from the forEach loops?

tinwelint replied:
Because the fusion instance is now stateful (closed flag) so they will all need to be recreated for each iteration

burqen reviewed (changes requested):
I think it became much more clear with the multi-target operations living inside `InstanceSelector`!

Still have a few comment, enjoy 🎉

Commit:
Extract lazy part of InstanceSelector into own class

To make it more explicit if instances are created lazily or not.
Hopefully making code easier to read and reason about.

tinwelint commented:
I like your latest changes 👍 @burqen

tinwelint commented:
@burqen all satisified here?

burqen commented:
@tinwelint Yep!

tinwelint committed:
License header updates
